### ISSUE TITLE
High RAM  using when  add (cache=True)in  Classification Training 

### ISSUE THREAD

**Issue by VyrionAI** [ultralytics/ultralytics#9824]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

when i add  (cache=True)in  Classification Training , the Ram using is increasing every epoch , until it crash the training , start like from 3   to 6  to 11 to 15 ....... 50 , GB
but if i don't add it , the ram using work fine , it be like 4 GB and all training is fixed 

i work on colab 
`!yolo task=classify mode=train cache=True model=yolov8n-cls.pt data='/content/Classification-1' epochs=5  batch=265 imgsz=128 `
  

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@VyrionAI hi there! 👋 It seems like you're encountering an issue with increased RAM usage when enabling `cache=True` during classification training. This is expected behavior as caching datasets in RAM speeds up training but requires more memory. 

Colab instances might not have enough memory for large datasets when caching is enabled. You might consider a few options:
1. Reduce the batch size to lower memory usage.
2. Use a smaller subset of your data.
3. Try training without caching if the dataset is too large for the available RAM. 

For example, to adjust your command:
```bash
!yolo task=classify mode=train cache=False model=yolov8n-cls.pt data='/content/Classification-1' epochs=5 batch=160 imgsz=128
```
Adjusting `batch` and experimenting with `cache=True/False` could help manage memory usage effectively. Remember, disabling cache will slow down training but prevent crashing due to out-of-memory errors. 

Hope this helps! Let us know if you have further questions. 😊

**Comment by glenn-jocher** [comment]
@VyrionAI Classify training with RAM cache currently has a bug where each worker caches separately, likely multiplying RAM requirements by 8x for 8x workers.

@Laughing-q is trying a fix in https://github.com/ultralytics/ultralytics/pull/9500 but there is no resolution likely soon.

Recommend cache=disk or cache=None for Classify training in the meantime.

**Comment by RMA2702** [comment]
Hello there, I encountered this issue while using yolo11 for classify training, the estimate training time would be so long if now training with RAM cache. Is there any resolution right now? Thanks!!

**Comment by Skillnoob** [comment]
@RMA2702 You can use disk caching. Currently ram caching has a bug as described above.

**Comment by RMA2702** [comment]
> @RMA2702 You can use disk caching. Currently ram caching has a bug as described above.

ok..thanks for the information

**Comment by glenn-jocher** [comment]
You're welcome! If you have any more questions, feel free to ask.

**Comment by longtran1904** [comment]
@glenn-jocher @Laughing-q I've recently digged deeper in Pytorch & YOLO caching and found this idea of Shared Memory Array

https://github.com/ptrblck/pytorch_misc/blob/master/shared_array.py

Wonder if it would help for this issue.

**Comment by glenn-jocher** [comment]
@longtran1904 thank you for the suggestion! We'll look into the Shared Memory Array approach to see if it can help address the caching issue.

**Comment by Laughing-q** [comment]
@longtran1904 Thanks for the sharing and insights! I tested the method but seems not really fit our current implementation of dataset since this share_array needs a fixed length while initializing, but we currently need to store images with its original shape which could be different across different images.
FYI current implementation of dataset pipeline in ultralytics package embeds the resize op into transforms/augmentations, but we need images stored before passing to transforms.

**Comment by WQL-call** [comment]
我为啥会报错WARNING ⚠️ Classification `cache_ram` training has known memory leak in https://github.com/ultralytics/ultralytics/issues/9824, setting `cache_ram=False`.
WARNING ⚠️ Classification `cache_ram` training has known memory leak in https://github.com/ultralytics/ultralytics/issues/9824, setting `cache_ram=False`.

**Comment by glenn-jocher** [comment]
The warning indicates that the `cache_ram` feature in classification training has a known memory leak issue, as detailed in [Ultralytics Issue #9824](https://github.com/ultralytics/ultralytics/issues/9824). To avoid this, the system automatically disables `cache_ram`. Using `cache=disk` or disabling caching entirely is recommended for classification tasks.

**Comment by WQL-call** [comment]
谢谢，非常感谢






------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "ultralytics/ultralytics"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年1月3日(星期五) 上午6:17
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [ultralytics/ultralytics] High RAM  using when  add (cache=True)in  Classification Training  (Issue #9824)





 
The warning indicates that the cache_ram feature in classification training has a known memory leak issue, as detailed in Ultralytics Issue #9824. To avoid this, the system automatically disables cache_ram. Using cache=disk or disabling caching entirely is recommended for classification tasks.

**Comment by glenn-jocher** [comment]
不客气！如果还有其他问题，请随时告诉我们。

**Comment by glenn-jocher** [comment]
@WQL-call 不客气！如果还有其他问题，请随时联系。感谢您对YOLO项目和Ultralytics的支持！

**Comment by WQL-call** [comment]
Dear Mr. Jocher,

I hope this message finds you well.

My name is Wang Qunli, and I am currently facing a challenge with my project where I have reached a plateau in terms of model accuracy improvements. As this is my first time utilizing YOLOv8, I have diligently reviewed the available documentation and performed numerous tests, yet I am still unable to achieve higher mAP (mean Average Precision) scores.

Given your expertise in this field, I would greatly appreciate any guidance or advice you could provide to help me address this issue and improve my model's performance.

Thank you very much for your time and consideration.

Best regards,

Wang Qunl






------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "ultralytics/ultralytics"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年1月6日(星期一) 晚上6:38
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [ultralytics/ultralytics] High RAM  using when  add (cache=True)in  Classification Training  (Issue #9824)





 
@WQL-call 不客气！如果还有其他问题，请随时联系。感谢您对YOLO项目和Ultralytics的支持！

**Comment by glenn-jocher** [comment]
Dear Wang Qunli,

Thank you for reaching out and sharing your challenge. To improve your YOLOv8 model's mAP, consider fine-tuning hyperparameters, augmenting your dataset, or using a larger model variant. Additionally, reviewing the [YOLOv8 documentation](https://docs.ultralytics.com/models/yolov8/) may offer further insights. Let us know if you encounter specific issues during the process.

**Comment by glenn-jocher** [comment]
Dear Wang Qunli,

Thank you for reaching out. To improve mAP with YOLOv8, consider strategies such as hyperparameter tuning, using a larger or more diverse dataset, applying advanced augmentations, or experimenting with different model sizes. For further guidance, the official [Ultralytics documentation](https://docs.ultralytics.com/) is an excellent resource. Unfortunately, we cannot provide personalized support but encourage you to share specific issues on the GitHub repository for community assistance.

**Comment by glenn-jocher** [comment]
Thanks for submitting PR [#19947](https://github.com/ultralytics/ultralytics/pull/19947) to address the RAM caching issue! This approach of main-thread preloading shows promise for resolving the per-worker duplication problem. We'll have the team review the implementation - @oystebw and core maintainers will provide feedback on the technical approach when they get a chance to evaluate it.

**Comment by glenn-jocher** [comment]
Verified current main `5eda95e83` beyond the merge status: #24670/#24673 replace the old per-worker lazy classification cache with preload plus the shared [dataset cache owner](https://github.com/ultralytics/ultralytics/blob/5eda95e839a9601fc479f1b5e042d48d7e77fa13/ultralytics/data/base.py), used by [ClassificationDataset](https://github.com/ultralytics/ultralytics/blob/5eda95e839a9601fc479f1b5e042d48d7e77fa13/ultralytics/data/dataset.py). Fresh real-image checks ran six complete training-loader passes with `cache=True`, two workers, and the official ImageNet10 training split (12 images), on macOS/spawn and Linux/fork. Worker RSS stabilized after startup: approximately **354.75/354.42 → 354.77/354.44 MiB** on macOS and **276.77/276.77 → 276.77/276.78 MiB** on Linux. The cached buffer hash stayed unchanged; explicit loader cleanup stopped both workers.

Closing the reported cache growth mechanism as fixed. These are small local lifecycle reproductions, not a 50-GB workload benchmark or a guarantee that a large RAM cache fits every machine. The contiguous NumPy buffer benefits from fork sharing; this is not a promise of zero-copy sharing under spawn. `cache=False` or disk caching remains appropriate when the cache does not fit. The separate current segmentation memory report #25832 and validation-pressure report #25827 remain open; these results do not establish their causes.
